rust/src/lib.rs

```diff
@@ -131,89 +132,116 @@ pub struct Allocation {
     }
-
     Ok(())
 }
 
-pub(crate) fn instrument_memory_grow<T>(
-    caller: Caller<T>,
+pub(crate) fn instrument_memory_grow(
     input: &[Val],
     _output: &mut [Val],
     ctx: Arc<Mutex<InstrumentationContext>>,
 ) -> Result<()> {
     let amount_in_pages = input[0].unwrap_i32(); // The number of pages requested by `memory.grow` instruction
-    let bt = WasmBacktrace::capture(caller);
-
-    // TODO: this should scream and die loudly if it can't actually get ahold of the frame
-    if let Some(frame) = bt.frames().last() {
-        if let Ok(mut cont) = ctx.lock() {
-            cont.allocate(frame, amount_in_pages as u32)?;
-        }
+    if let Ok(mut cont) = ctx.lock() {
+        cont.allocate(amount_in_pages as u32)?;
     }
     Ok(())
 }
 
 const MODULE_NAME: &str = "dylibso_observe";
 
 type EventChannel = (Sender<Event>, Receiver<Event>);
+type FunctionNames = HashMap<u32, String>;
 
 /// Link observability import functions required by instrumented wasm code
-pub fn add_to_linker<T: 'static>(id: usize, linker: &mut Linker<T>) -> Result<EventChannel> {
+pub fn add_to_linker<T: 'static>(id: usize, linker: &mut Linker<T>, data : &Vec<u8>) -> Result<EventChannel> {
     let (ctx, events_tx, events_rx) = InstrumentationContext::new(id);
 
-    let t = FuncType::new([], []);
+    // Build up a table of function id to name mapping
+    let mut function_names = FunctionNames::new();
+    let parser = wasmparser::Parser::new(0);
+    for payload in parser.parse_all(&data) {
+        match payload? {
+            wasmparser::Payload::CustomSection(custom) => {
+                if custom.name() == "name" {
+                    let name_reader =
+                        wasmparser::NameSectionReader::new(custom.data(), custom.data_offset());
+                    for x in name_reader.into_iter() {
+                        let x = x?;
+                        match x {
+                            //wasmparser::Name::Module { name, name_range } => (),
+                            wasmparser::Name::Function(f) => {
+                                for k in f.into_iter() {
+                                    let k = k?;
+                                    function_names.insert(k.index, k.name.to_string());
+                                }
+                            },
+                            _ => ()
+                            //wasmparser::Name::Local(_) => (),
+                            //wasmparser::Name::Label(_) => (),
+                            //wasmparser::Name::Type(_) => (),
+                            //wasmparser::Name::Table(_) => (),
+                            //wasmparser::Name::Memory(_) => (),
+                            //wasmparser::Name::Global(_) => (),
+                            //wasmparser::Name::Element(_) => (),
+                            //wasmparser::Name::Data(_) => (),
+                            //wasmparser::Name::Unknown { ty, data, range } => (),
+                        }
```

Comment:
If you just care about matching one item, you can use `if let`. Something like this:

```rust
if let wasmparser::Name::Function(f) = x? {

}
```

Same could apply for the above match statement.

Not a big deal but often flows a little better when you have a match with one arm: https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

Reply:
Done, good idea too!